### PR TITLE
Fix memory leak due to misuse of K8s clients

### PR DESCRIPTION
**Pull request by misberner**
The existing code creates two(!) Kubernetes client factories for each(!) reconciliation of each(!) object. The factories do get garbage collected, but because of goroutines, this does not happen immediately. Those factories accumulate a lot of memory, mostly due to the parsing of discovery data/server OpenAPI schema:
Screen Shot 2022-09-08 at 02 18 07
In fact, by inserting an error into the CR (and thereby causing a hot reconciliation loop), I could make the memory go above 2G in less than a minute.

Rewrite the code such that a single Kubernetes client factory is instantiated (through `kube.New`, factory being reused for creating a clientset for the storage driver secrets client), and also a single REST client getter. With this change, the memory of the ACS operator remains at ~80MB, even when the CR contains an error and there is little to no back-off between reconciles.

### CODE REVIEW

File: pkg/client/actionconfig.go [hunk 1]
```diff
@@ -30,72 +32,76 @@ import (
...
 	ActionConfigFor(obj client.Object) (*action.Configuration, error)
 }
 
-func NewActionConfigGetter(cfg *rest.Config, rm meta.RESTMapper, log logr.Logger) ActionConfigGetter {
-	return &actionConfigGetter{
-		cfg:        cfg,
-		restMapper: rm,
-		log:        log,
-	}
+type errActionConfigGetter struct {
```
Comment (SimonBaeumer) [inline pkg/client/actionconfig.go]: Can you add a comment why an `errActionConfigGetter` is introduced?
It may not be obvious to a reader to understand the reasoning that this prevents an OOM crash in cases where lots of reconciles happens like errors.

File: pkg/client/actionconfig.go [hunk 2]
```diff
@@ -30,72 +32,76 @@ import (
...
-	// Create a client that helm will use to manage release resources.
-	// The passed object is used as an owner reference on every
-	// object the client creates.
 	kc := kube.New(rcg)
 	kc.Log = debugLog
 
-	// Create the Kubernetes Secrets client. The passed object is
-	// also used as an owner reference in the release secrets
-	// created by this client.
-	kcs, err := cmdutil.NewFactory(rcg).KubernetesClientSet()
```
Comment (SimonBaeumer) [inline pkg/client/actionconfig.go:77]: Could you explain why this change is possible and still setting the owner ref on the release secret?

**Review by SimonBaeumer — CHANGES_REQUESTED** [review]
Great @misberner!

File: pkg/client/actionconfig.go (hunk 2 above)
Comment (misberner) [reply pkg/client/actionconfig.go:77]: Because the secrets client is still wrapped in line 86 (new) below

File: pkg/client/actionconfig.go (hunk 1 above)
Comment (misberner) [reply pkg/client/actionconfig.go]: Well, actually I removed the `errActionConfigGetter`, and the reasoning to introduce it was not to prevent OOMs (the other fixes accomplish that).

The reason I introduced an `errActionConfigGetter` was that it seemed inconvenient to change the signature of this function to return an `error` in addition to the `ActionConfigGetter`, especially due to it being used in tests. However, after I gave it a second look it turned out to be not that big of a deal, so I changed the signature to return an `error` and no longer need an `errActionConfigGetter`.

**Review by SimonBaeumer — APPROVED** [review]

File: pkg/client/actionconfig.go [hunk 3]
```diff
@@ -30,72 +32,68 @@ import (
...
 
 	// Also, use the debug log for the storage driver
-	d.Log = debugLog
+	d.Log = acg.debugLog
 
 	// Initialize the storage backend
 	s := storage.Init(d)
 
+	kubeClient := *acg.kubeClient
+	kubeClient.Namespace = obj.GetNamespace()
```
Comment (porridge) [inline pkg/client/actionconfig.go:90]: @misberner perhaps I'm missing something, but this seems not to be thread-safe. I seem to recall that a reconciler can be configured to run multiple reconciles in parallel, in which case the returned `action.Configuration` objects could used concurrently, and only the `.KubeClient` of last one would have the correct `.Namespace` set, no?

**Review by porridge — COMMENTED** [review]
Did you perhaps run this change by @joelanford already?

**Comment by misberner** [comment]
> Did you perhaps run this change by @joelanford already?

Yes, I've sent out an upstream PR.

File: pkg/client/actionconfig.go (hunk 3 above)
Comment (misberner) [reply pkg/client/actionconfig.go:90]: The `:= *...` creates a shallow copy, which is sufficient for thread-safety since `Namespace` is a top-level field and therefore not aliased between copies.
